row_builder.go

```diff
@@ -0,0 +1,185 @@
+}
+
+func (b *RowBuilder) configure(node Node, columnIndex int16, level columnLevel, group *columnGroup) (endIndex int16) {
+	switch {
+	case node.Optional():
+		level.definitionLevel++
+		endIndex = b.configure(Required(node), columnIndex, level, group)
+
+		for i := columnIndex; i < endIndex; i++ {
+			b.kinds[i] = 0 // null if not set
+		}
+
+	case node.Repeated():
+		level.definitionLevel++
+
+		group = &columnGroup{
+			startIndex:      columnIndex,
+			repetitionLevel: level.repetitionDepth,
+			definitionLevel: level.definitionLevel,
+		}
+
+		level.repetitionDepth++
+		endIndex = b.configure(Required(node), columnIndex, level, group)
+
+		for i := columnIndex; i < endIndex; i++ {
+			b.kinds[i] = 0 // null if not set
+		}
+
+		group.endIndex = endIndex
+		b.groups = append(b.groups, group)
+
+	case node.Leaf():
+		group.members = append(group.members, columnIndex)
+		b.kinds[columnIndex] = ^int8(node.Type().Kind())
+		b.levels[columnIndex] = level
+		endIndex = columnIndex + 1
+
+	default:
+		endIndex = columnIndex
+
+		for _, field := range node.Fields() {
+			endIndex = b.configure(field, endIndex, level, group)
+		}
+	}
+	return endIndex
+}
+
+// Add adds columnValue to the column at columnIndex.
+func (b *RowBuilder) Add(columnIndex int, columnValue Value) {
+	level := &b.levels[columnIndex]
+	columnValue.repetitionLevel = level.repetitionLevel
+	columnValue.definitionLevel = level.definitionLevel
+	columnValue.columnIndex = ^int16(columnIndex)
+	level.repetitionLevel = level.repetitionDepth
+	b.columns[columnIndex] = append(b.columns[columnIndex], columnValue)
+}
+
+// Next must be called to indicate the start of a new repeated record for the
+// column at the given index.
+//
+// If the column index is part of a repeated group, the builder automatically
+// starts a new record for all adjacent columns, the application does not need
+// to call this method for each column of the repeated group.
+//
+// Next must be called after adding a sequence of records.
+func (b *RowBuilder) Next(columnIndex int) {
+	for _, group := range b.groups {
+		if group.startIndex <= int16(columnIndex) && int16(columnIndex) < group.endIndex {
+			for i := group.startIndex; i < group.endIndex; i++ {
+				level := &b.levels[i]
+				level.repetitionLevel = group.repetitionLevel
+			}
+			break
+		}
+	}
+}
+
+// Reset clears the internal state of b, making it possible to reuse while
+// retaining the internal buffers.
+func (b *RowBuilder) Reset() {
+	for i, column := range b.columns {
+		clearValues(column)
+		b.columns[i] = column[:0]
+	}
+	for i := range b.levels {
+		b.levels[i].repetitionLevel = 0
+	}
+}
+
+// Row materializes the current state of b into a parquet row.
+func (b *RowBuilder) Row() Row {
+	numValues := 0
+	for _, column := range b.columns {
+		numValues += len(column)
+	}
+	return b.AppendRow(make(Row, 0, numValues))
+}
+
+// AppendRow appends the current state of b to row and returns it.
+func (b *RowBuilder) AppendRow(row Row) Row {
```

Comment:
FWIW, in the standard library, the Append() functions are static functions that take an existing slice and new value and then return the new slice. Ie here it would be something like

```
func AppendRow(row Row, newContents interface{}(?)) Row

Reply:
🤔 there is [`time.(Time).AppendFormat`](https://pkg.go.dev/time#Time.AppendFormat)

I don't think there's a strong convention being enforced in the stdlib here, but maybe I just don't have enough context on the design discussions that are going on behind the scenes.